misk-hibernate/src/main/kotlin/misk/hibernate/InvalidChildEntityLoadException.kt

```diff
@@ -0,0 +1,13 @@
+package misk.hibernate
+
+import org.hibernate.TypeMismatchException
+
+/**
+ * [InvalidChildEntityLoadException] is raised when there is an attempt to retrieve a child entity
+ * using an id. Naturally, loading a child by id without a lookup will result in a scatter query,
+ * which we want to avoid. Hibernate will catch these with a `TypeMismatchException` as the entity
+ * will expect a gid, but we'll rewrite those as `InvalidChildEntityLoadException` for clarity. The
+ * expected way to load child entities is to use `loadByGid()`
+ */
+class InvalidChildEntityLoadException(e: TypeMismatchException) : Exception(e) {
```

Comment:
I think we can probably get away without this? Fewer exception types is preferable, unless there’s somebody who wants to catch one exception and not another

Reply:
I think the main incentive here is clarity - something like `session.load(entityId)` feels pretty natural to do, but will blow up if you use a child entity. Rather than seeing `TypeMismatchException: Expected gid but got id` and not necessarily understanding the context around the exception - I think highlighting the context of `InvalidChildEntityLoad` makes it clear where the error is rooted. Though I suppose I eventually found my way, so I guess other people could too? Not too attached to this idea - but thought it could be useful.

Reply:
I agree. We definitely want the errors to be entirely self explanatory and `TypeMismatchException` is not. Adding an extra exception here with a nice explanatory JavaDoc is a big win for engineers self sufficiency. Self sufficiency increases independence and independence increases job satisfaction.